diff --git a/pynq/lib/video/clocks.py b/pynq/lib/video/clocks.py
index 609f5bc4..92fcf777 100644
--- a/pynq/lib/video/clocks.py
+++ b/pynq/lib/video/clocks.py
@@ -6,7 +6,8 @@ import cffi
 import math
 import numpy as np
 from .constants import *
-
+from pynq import GPIO
+import time
 
 _ffi = cffi.FFI()
 
@@ -737,3 +738,152 @@ class SI_5324C:
         return result
 
 
+# The following algorithm is transcribed from the TySOM-3A-ZU19EG HDMI reference design
+
+SI570_XTAL_FREQ = 40000000   # The freq of the crystal in Hz
+SI570_FDCO_MIN = 4850000000  # Min VCO Operating Freq in Hz
+SI570_FDCO_MAX = 5670000000  # Max VCO Operating Freq in Hz
+
+SI570_FOUT_MIN = 10000000    # Min Output Freq in Hz
+SI570_FOUT_MAX = 945000000   # Max Output Freq in Hz
+
+SI570_P_MAX = 4194304        # pow(2,22) - Max P div value
+SI570_M_MAX = 16777216       # pow(2,24) - Max M mult value
+
+HS_DIV_OFFSET = 4
+SI570_FACTORY_FREQ = 156250000
+
+class SI570:
+    """Driver for the Si570 clock generator
+
+    """
+
+    def __init__(self, master, address, gpio):
+        """Create a new instance of the Si570 driver
+
+        Parameters
+        ----------
+        master : IIC master
+            IIC master the device is connected to
+        address : int
+            IIC address of the device
+
+        """
+        self._master = master
+        self._address = address
+        self._buffer = _ffi.new("unsigned char [32]")
+        self._rdy = GPIO(GPIO.get_gpio_pin(gpio), 'out')
+
+    def _read(self, reg_addr):
+        self._buffer[0] = reg_addr
+        self._master.send(self._address, self._buffer, 1, 1)
+        self._master.receive(self._address, self._buffer, 1)
+        self._master.wait()
+        # Clear all of the interrupts
+        self._master.write(0x20, self._master.read(0x20))
+        return self._buffer[0]
+
+    def _write(self, reg_addr, data):
+        self._buffer[0] = reg_addr
+        self._buffer[1] = data
+        self._master.send(self._address, self._buffer, 2)
+        self._master.wait()
+        # Clear all of the interrupts
+        self._master.write(0x20, self._master.read(0x20))
+
+    def _update(self, reg_addr, value, mask):
+        data = self._read(reg_addr)
+        data &= ~mask
+        data |= (value & mask)
+        self._write(reg_addr, data)
+
+    def read_rfreq(self):
+        rfreq = (self._read(0x0008) & 0x3F) << 32
+        rfreq |= self._read(0x0009) << 24
+        rfreq |= self._read(0x000A) << 16
+        rfreq |= self._read(0x000B) << 8
+        rfreq |= self._read(0x000C)
+        return rfreq
+
+    def set_rfreq(self, value):
+        self._update(0x0008, (value >> 32) & 0x3F, 0x3F)
+        self._write(0x0009, (value >> 24) & 0xFF)
+        self._write(0x000A, (value >> 16) & 0xFF)
+        self._write(0x000B, (value >> 8) & 0xFF)
+        self._write(0x000C, (value & 0xFF))
+
+    def read_hsdiv(self):
+        hsdiv = ((self._read(0x0007) & 0xE0) >> 5) + HS_DIV_OFFSET
+        return hsdiv
+
+    def set_hsdiv(self, value):
+        self._update(0x0007, ((value - HS_DIV_OFFSET) << 5) & 0x3F, 0x3F)
+
+    def read_n1(self):
+        n1 = (self._read(0x0007) & 0x1F) << 2
+        n1 |= (self._read(0x0008) & 0xC0) >> 6
+        n1 += 1
+        return n1
+
+    def set_n1(self, value):
+        self._update(0x0007, ((value - 1) >> 2) & 0x1F, 0x1F)
+        self._update(0x0008, ((value - 1) << 6) & 0xC0, 0xC0)
+
+    def freeze(self, active):
+        if active:
+            value = 0x10
+        else:
+            value = 0x00
+        self._update(0x0089, value, 0x10)
+
+    def recall(self, active):
+        if active:
+            value = 0x1
+        else:
+            value = 0x0
+        self._update(0x087, value, 0x1)
+
+    def new_freq(self):
+        self._update(0x085, 1 << 6, 1 << 6)
+
+    def set_clock(self, freq, line_rate):
+        self._set_clock(freq)
+
+    def _set_clock(self, fout):
+        if fout < SI570_FOUT_MIN:
+            raise RuntimeError("Output Frequency Below Minimum")
+        if fout > SI570_FOUT_MAX:
+            raise RuntimeError("Output Frequency Above Maximum")
+
+        fdco_min = SI570_FDCO_MAX
+
+        #settings = _calculate_settings(fin, fout)
+        self.recall(True)
+        time.sleep(0.015)
+        rfreq = self.read_rfreq()
+        hsdiv = self.read_hsdiv()
+        n1 = self.read_n1()
+
+        # Calculate the actual nominal crystal frequency
+        f_xtal = (SI570_FACTORY_FREQ * hsdiv * n1) / rfreq * 2**28
+
+        # Choose the output dividers
+        for i in range(4, 12):
+            for j in range(2, 128, 2):
+                fdco = fout * i * j
+                if(fdco > SI570_FDCO_MIN and fdco < SI570_FDCO_MAX):
+                    rfreq = fdco / f_xtal
+                    rfreq = int(rfreq * 2**28)
+                    if(fdco < fdco_min):
+                        fdco_min = fdco
+                        best_hs_div = i
+                        best_n1 = j
+
+        self._rdy.write(1)
+        self.freeze(False)
+        self.set_rfreq(rfreq)
+        self.set_hsdiv(best_hs_div)
+        self.set_n1(best_n1)
+        self.new_freq()
+        self.freeze(True)
+        self._rdy.write(0)
diff --git a/sdbuild/Makefile b/sdbuild/Makefile
index aadf3583..1b1abb5b 100644
--- a/sdbuild/Makefile
+++ b/sdbuild/Makefile
@@ -159,7 +159,7 @@ $$(PL_PROJ_$1): $$(BSP_TARGET_$1)
 		-s $$(BSP_BUILD_$1)/$$(BSP_PROJECT_$1).bsp -n petalinux_project
 	echo 'CONFIG_USER_LAYER_0="'$(CURDIR)/boot/meta-pynq'"' >> $$(PL_CONFIG_$1)
 	echo 'CONFIG_SUBSYSTEM_ROOTFS_EXT4=y' >> $$(PL_CONFIG_$1)
-	echo 'CONFIG_SUBSYSTEM_SDROOT_DEV="/dev/mmcblk0p2"' >> $$(PL_CONFIG_$1)
+	echo 'CONFIG_SUBSYSTEM_SDROOT_DEV="/dev/mmcblk1p2"' >> $$(PL_CONFIG_$1)
 	echo 'CONFIG_SUBSYSTEM_ETHERNET_MANUAL_SELECT=y' >> $$(PL_CONFIG_$1)
 	if [ $$(FPGA_MANAGER_$1) = 1 ]; then \
 		echo 'CONFIG_SUBSYSTEM_DEVICETREE_FLAGS="-@"' >> $$(PL_CONFIG_$1) ;\
diff --git a/sdbuild/boot/meta-pynq/recipes-bsp/device-tree/files/pynq_bootargs.dtsi b/sdbuild/boot/meta-pynq/recipes-bsp/device-tree/files/pynq_bootargs.dtsi
index f9964316..60275fe6 100644
--- a/sdbuild/boot/meta-pynq/recipes-bsp/device-tree/files/pynq_bootargs.dtsi
+++ b/sdbuild/boot/meta-pynq/recipes-bsp/device-tree/files/pynq_bootargs.dtsi
@@ -1,6 +1,6 @@
 / {
 	chosen {
-		bootargs = "root=/dev/mmcblk0p2 rw earlyprintk rootfstype=ext4 rootwait devtmpfs.mount=1 uio_pdrv_genirq.of_id=\"generic-uio\" clk_ignore_unused";
+		bootargs = "root=/dev/mmcblk1p2 rw earlyprintk rootfstype=ext4 rootwait devtmpfs.mount=1 uio_pdrv_genirq.of_id=\"generic-uio\" clk_ignore_unused";
 		pynq_board = "Unknown";
 	};
 };
diff --git a/sdbuild/packages/bootpy/bootpy.sh b/sdbuild/packages/bootpy/bootpy.sh
index f8b7f738..521f3f7e 100755
--- a/sdbuild/packages/bootpy/bootpy.sh
+++ b/sdbuild/packages/bootpy/bootpy.sh
@@ -4,7 +4,7 @@
 for f in /etc/profile.d/*.sh; do source $f; done
 
 BOOT_MNT=/boot
-BOOT_DEV=/dev/mmcblk0p1
+BOOT_DEV=/dev/mmcblk1p1
 BOOT_PY=$BOOT_MNT/boot.py
 
 if !(mount | grep -q "$BOOT_MNT") ; then
diff --git a/sdbuild/packages/resizefs/resizefs.sh b/sdbuild/packages/resizefs/resizefs.sh
index a60700cf..0b92e40d 100755
--- a/sdbuild/packages/resizefs/resizefs.sh
+++ b/sdbuild/packages/resizefs/resizefs.sh
@@ -8,8 +8,8 @@
 exec >> /var/log/syslog
 exec 2>&1
 
-TGTDEV=/dev/mmcblk0
-TGTPART=/dev/mmcblk0p2
+TGTDEV=/dev/mmcblk1
+TGTPART=/dev/mmcblk1p2
 source /etc/environment
 
 if [[ ${RESIZED} -eq "1" ]]; then
diff --git a/sdbuild/packages/wifi_connect/wifi_connect.sh b/sdbuild/packages/wifi_connect/wifi_connect.sh
index b7ff24a3..5e80f2bf 100755
--- a/sdbuild/packages/wifi_connect/wifi_connect.sh
+++ b/sdbuild/packages/wifi_connect/wifi_connect.sh
@@ -7,7 +7,7 @@ config_name="wpa_supplicant.conf"
 driver_name="wifi.ko"
 
 #use the 1st partition of sdcard
-boot_part="/dev/mmcblk0p1"
+boot_part="/dev/mmcblk1p1"
 
 #mount boot partition to temp folder
 mount -o rw "$boot_part" "$mount"
diff --git a/sdbuild/boot/image.wks b/sdbuild/boot/image.wks
index 8ca728b6..35fa456b 100644
--- a/sdbuild/boot/image.wks
+++ b/sdbuild/boot/image.wks
@@ -1,6 +1,6 @@
 # Description: Creates a partitioned SD card image. Boot files
 # are located in the first vfat partition. Rootfs will be in second ext4 partition.
 
-part /boot --source bootimg-partition --ondisk mmcblk0 --fstype=vfat --label PYNQ --active --align 4 --size 100 --overhead-factor=1
-part / --source rootfs --ondisk mmcblk0 --fstype=ext4 --label root --align 1024 --extra-space 512
+part /boot --source bootimg-partition --ondisk mmcblk1 --fstype=vfat --label PYNQ --active --align 4 --size 100 --overhead-factor=1
+part / --source rootfs --ondisk mmcblk1 --fstype=ext4 --label root --align 1024 --extra-space 512
 
